crates/bevy_window/src/lib.rs

```diff
@@ -78,27 +91,56 @@ impl Plugin for WindowPlugin {
             .add_event::<WindowScaleFactorChanged>()
             .add_event::<WindowBackendScaleFactorChanged>()
             .add_event::<FileDragAndDrop>()
-            .add_event::<WindowMoved>()
-            .init_resource::<Windows>();
-
-        if self.add_primary_window {
-            app.world.send_event(CreateWindow {
-                id: WindowId::primary(),
-                descriptor: self.window.clone(),
-            });
+            .add_event::<WindowMoved>();
+
+        if let Some(initial_window_bundle) = &self.initial_window {
+            // If the user has added a window-bundle resource, we should spawn that as as the
+            // initial window. If not, we need to spawn a default WindowBundle,
+            // hence the `unwrap_or_default()` here.
+            let window_id = app.world.spawn(initial_window_bundle.clone()).id();
+
+            bevy_utils::tracing::info!("Initial window: {:?}", window_id);
+
+            app.world
+                .insert_resource(InitialWindow { window: window_id });
         }
 
-        if self.exit_on_all_closed {
-            app.add_system_to_stage(
-                CoreStage::PostUpdate,
-                exit_on_all_closed.after(ModifiesWindows),
-            );
+        match self.exit_condition {
+            ExitCondition::OnInitialClosed => {
+                app.add_system(exit_on_initial_closed);
+            }
+            ExitCondition::OnAllClosed => {
+                app.add_system(exit_on_all_closed);
+            }
+            ExitCondition::DontExit => {}
         }
+
         if self.close_when_requested {
-            app.add_system(close_when_requested);
+            app.add_system_to_stage(CoreStage::First, close_when_requested);
         }
     }
 }
 
+/// System Label marking when changes are applied to windows
 #[derive(Debug, Hash, PartialEq, Eq, Clone, SystemLabel)]
 pub struct ModifiesWindows;
+
+/// Defines the specific conditions the application should exit on
+#[derive(Clone)]
+pub enum ExitCondition {
+    /// Close application when the initial window is closed
+    OnInitialClosed,
+    /// Close application when all windows are closed
+    OnAllClosed,
+    /// Keep application running headless even after closing all windows
+    DontExit,
+}
+
+/// The window that was created on startup.
+///
+/// This resource is allowed to not exist and should be handled gracefully if it doesn't.
```

Comment:
This comment should explain what's used when not.

Reply:
replaced with marker component